docs/rfcs/20220131_raft_snapshots.md

```diff
@@ -0,0 +1,149 @@
+- Feature Name: Raft snapshots
+- Status: draft
+- Start Date: 2022-01-31
+- Authors: Denis Rystsov
+- Issue: TBD
+
+# Executive Summary
+
+Introduce Raft snapshots not bound to log truncation to solve compaction related problems with idempotency, transactions, offset translator and pave a way to custom state machines.
+
+## What is being proposed
+
+Make the snapshot before the compaction and ship it to all (majority) nodes in a cluster so that each node sees the same aggregate of the log. Currently when Redpanda does compaction it introduces holes to the log and when a follower replay the log to build an aggregate it end up with different view.
+
+# Motivation
+## Why are we doing this?
+
+To fix existing problems which prevents transacitons from working with the compacted topics. To introduce infastructure to support custom (probably WASM based) state machines.
+
+## What use cases does it support?
+
+ - Idempotency
+ - Transactions
+ - Offset translator
+ - Custom state machines
+
+# Guide-level explanation
+## Context
+
+Some of the Redpanda operations require knowledge about all records passed through the partition (Raft log) up to this moment (let's call them aggregate-dependent).
+ 
+ - Idempotency: Redpanda inspects the messages to update the list of ongoing sessions and then uses it to reject the out of order requests.
+ - Offset translator: Redpanda tracks the reconfiguration batches to adjust raft offsets into Kafka offsets.
+ - Transactions: Redpanda inspects the records to keep an eye on the ongoing transactions and maintain the right consistency level
+ 
+At the same time disks have limited capacity and data eventually lose actuality so Redpanda supports several cleanup policies:
+ 
+ - Deletion
+ - Compaction
+ 
+Deletion clean up policy removes some messages (defined by age or size) from the head of the log.
+
+![](./imgs/20220131_log.png)
+
+Compact clean up policy is a bit more complicated: each record in Redpanda has a key and with compaction enabled Redpanda keeps a record at the tail of the log but removes some or all records with the same key written before it.
+
+## Problem
+
+There is a problem with the integration of aggregate-dependent operations and the cleanup policies: Redpanda may remove data before the operation sees it (it's especially true during reconfiguration because the new node never saw a non compacted log).
```

Comment:
```suggestion
There is a problem with the integration of aggregate-dependent operations and the cleanup policies: Redpanda may remove data before the operation sees it. This is especially possible during reconfiguration because the new node has never seen a non-compacted log).
```